### PR TITLE
docs: Introduce TypeDoc auto-generated API docs at /docs/typedoc

### PR DESCRIPTION
**Pull request by kamilogorek**
👀 👀 👀 

https://user-images.githubusercontent.com/1523305/221823129-ec9c6a59-a897-4f18-a375-a606f9c4f1f4.mp4

Closes https://github.com/trpc/trpc/issues/3655

## 🎯 Changes

This change introduces a new position into the documentation sidebar called `TypeDoc`, which is reachable via `/docs/typedoc` path.

Docs are generated automatically during the build process.

`docusaurus-plugin-typedoc` and `typedoc-plugin-markdown` are using `next` versions, as there are plenty of improvements to rendering and available compiler options, that it's worth it to "live on the edge".

Tbh it was quite a painful process, as it was not obvious whether something should be configured in typedoc, docusaurus, plugin-markdown, or plugin-typedoc, and at the same time some option names are different here and there, nor they are synced and compatible with each other, but I somehow got it all married together.

Magic, magic everywhere.

(also that pnpm lockfile changing on every large PR is quite annoying 😬)

## ✅ Checklist

- [x] I have followed the steps listed in the [Contributing guide](https://github.com/trpc/trpc/blob/main/CONTRIBUTING.md).
- [x] If necessary, I have added documentation related to the changes made.
- [x] I have added or updated the tests related to the changes made.

### CODE REVIEW

File: www/docusaurus.typedoc.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
+function generateTypedocDocusaurusPlugins(directories) {
+  return directories.map((directory, idx) => [
+    'docusaurus-plugin-typedoc',
+    {
+      id: directory,
+      entryPoints: [`../packages/${directory}/src/index.ts`],
+      tsconfig: `../packages/${directory}/tsconfig.build.json`,
+      out: `./typedoc/${directory}`,
```
Comment (juliusmarminge) [inline www/docusaurus.typedoc.js:8]: this output must be added to the turbo pipeline

Comment (juliusmarminge) [reply www/docusaurus.typedoc.js:8]: this is probably why there is only react and client package in the deployed version:

CleanShot 2023-02-28 at 11 17 56@2x

turbo didn't see any changes to server or next so probably skipped it or something?

Comment (kamilogorek) [reply www/docusaurus.typedoc.js:8]: Do you know what commands exactly are run during deployment? Is is possible to reproduce the same build env locally?

Comment (juliusmarminge) [reply www/docusaurus.typedoc.js:8]: same command. if you remove the build & typedoc folders and rerun build-www, you should hit the turbo cache and it'll restore the build output but not typedoc (since you haven't told it that it outputs that)

Comment (juliusmarminge) [reply www/docusaurus.typedoc.js:8]: Run in root to build with turbo:

```sh
turbo --filter www build
```

**Comment by KATT** [comment]
Wow, this is super cool

File: www/docusaurus.typedoc.js (hunk 1 above)
Comment (kamilogorek) [reply www/docusaurus.typedoc.js:8]: That's quite odd, as I verified everything just now, and it appears that turbo caches contain all necessary files.

image

Let's see if adding this generated content to the pipeline will do the job.

Comment (kamilogorek) [reply www/docusaurus.typedoc.js:8]: It did not 👀 Could someone maybe check the Vercel logs whether `Documentation generated at ./docs/typedoc` is correctly executed for all 4 packages?

Comment (juliusmarminge) [reply www/docusaurus.typedoc.js:8]: it's not in the build output

CleanShot 2023-02-28 at 11 58 21@2x

Comment (juliusmarminge) [reply www/docusaurus.typedoc.js:8]: Strange... 
CleanShot 2023-02-28 at 12 00 33@2x

Comment (kamilogorek) [reply www/docusaurus.typedoc.js:8]: Ah, they are not generated due to build errors, as typedoc will skip those.

Comment (kamilogorek) [reply www/docusaurus.typedoc.js:8]: Rebased on top of `main`, as there were some changes to `pnpm-lock` in the meantime. Lets hope it helps.

**Comment by kamilogorek** [comment]
Fixed deployment issue. What happened is that Vercel was running
```
pnpm --filter root --filter www --filter '@trpc/*' --filter '!@trpc/tests' install
```
which effectively prevented any `tests` dependencies to be installed. Which _is_ indeed correct.

This however exposed the issue with `@types/react-dom` transitive dependency, which was "accidentally" installed via one of `tests` package dependencies, so it never blew up and explicitly told the user that there was an issue with unresolved dependency.

TypeDoc however, is using `tsconfig.build.json`, which will perform a full packages build, and effectively will require all necessary dependencies in isolation. This exposed the issue of missing dep, which has been now fixed.

**Comment by ixahmedxi** [comment]
Do you not think `API Reference` is a better name for this instead of `TypeDoc`? I see many implementations of this using that section name and for people that don't know what `TypeDoc` is it will sound more familiar. Looks amazing tho I love it, great work <3

**Comment by kamilogorek** [comment]
It's just a label, so it won't affect the URL itself though. I have no strong feelings about either of those, so you can internally decide what should be the url scheme and what should be the naming convention here, and I'll update it accordingly :)

It's also possible to create custom readme pages for each and still preserve generated table of content, but I'd say its a nice enhancement for the future if there'll be a need for one.

Thanks!

**Comment by kamilogorek** [comment]
Flaky E2E tests are not related to this change, so it should be ready to merge :)

**Comment by kamilogorek** [comment]
Added `**/www/typedoc/` to `.prettierignore`, rebased on top of `main` and cleaned up the PR.

File: packages/tests/server/interop/inference.test.ts [hunk 2]
```diff
@@ -204,7 +204,7 @@ describe('infer subscription input & output', () => {
       },
     })
     .interop();
-  type TSubscriptions = typeof router['_def']['subscriptions'];
+  type TSubscriptions = (typeof router)['_def']['subscriptions'];
```
Comment (KATT) [inline packages/tests/server/interop/inference.test.ts:207]: Where are these changes coming from? Updated prettier or lint?

**Comment by kamilogorek** [comment]
Answered in the comment above :)

> Parentheses changes are coming from prettier@2.8.2 - https://github.com/prettier/prettier/blob/main/CHANGELOG.md#add-parentheses-for-tstypequery-to-improve-readability-14042-by-onishi-kohei

File: packages/tests/server/interop/inference.test.ts (hunk 2 above)
Comment (kamilogorek) [reply packages/tests/server/interop/inference.test.ts:207]: Answered in the comment above :)

> Parentheses changes are coming from prettier@2.8.2 - https://github.com/prettier/prettier/blob/main/CHANGELOG.md#add-parentheses-for-tstypequery-to-improve-readability-14042-by-onishi-kohei

File: README.md [hunk 3]
```diff
@@ -258,6 +259,6 @@ Also, if your company using tRPC and want to support long-term maintenance of tR
 
 <a href="https://vercel.com/?utm_source=trpc&utm_campaign=oss">
   <p align="center">
-    <img src="./images/powered-by-vercel.svg" alt="Powered by Vercel" title="Powered by Vercel">
+    <img src="./images/powered-by-vercel.svg" alt="Powered by Vercel" title="Powered by Vercel" />
```
Comment (KATT) [inline README.md:262]: unrelated
```suggestion
    <img src="./images/powered-by-vercel.svg" alt="Powered by Vercel" title="Powered by Vercel">
```

File: README.md [hunk 4]
```diff
@@ -9,7 +9,7 @@
   <h1>tRPC</h1>
   <h3>Move fast and break nothing.<br />End-to-end typesafe APIs made easy.</h3>
   <a href="https://codecov.io/gh/trpc/trpc">
-    <img alt="codecov" src="https://codecov.io/gh/trpc/trpc/branch/main/graph/badge.svg?token=KPPS918B0G">
+    <img alt="codecov" src="https://codecov.io/gh/trpc/trpc/branch/main/graph/badge.svg?token=KPPS918B0G" />
```
Comment (KATT) [inline README.md:12]:
```suggestion
    <img alt="codecov" src="https://codecov.io/gh/trpc/trpc/branch/main/graph/badge.svg?token=KPPS918B0G">
```

File: README.md [hunk 5]
```diff
@@ -92,6 +92,7 @@ npx create-next-app --example https://github.com/trpc/trpc --example-path exampl
 
 <table>
   <tbody>
+    <tr>
```
Comment (KATT) [inline README.md:95]: ok i'll keep this haha

**Review by KATT — DISMISSED** [review]

**Comment by kamilogorek** [comment]
Or well, maybe not, and www failures are completely due to different reason now. Impossible to tell without logs.
